### PR TITLE
Remove hardcode specific to Clinical Genomics

### PR DESCRIPTION
**Pull request by adrosenbaum**
User can now specify in a config file what data to be exported to
the vcf file from the database.

This PR also includes some refactoring
adding integration test to travis

black formatting for touched files

This PR looks immense! and yes there are quite a few changes. However, most of the changed code are black formatted lines, and refactoring...

**How to test**:
1. install on stage `bash /home/proj/stage/servers/resources/hasta.scilifelab.se/updater-mutacc-stage.sh remove-hardcode`
1. activate stage: `us`
1. run following command: `cg upload process-solved --case-id <case_id>`
1. `cg upload processed-solved>`
1. `mutacc db export --all-variants -m child -n child -p`

**Expected outcome**:
exported vcf should contain info from the database that is specified in the config file

**Review:**
- [x] code approved by
- [x] tests executed by
- [ ] "Merge and deploy" approved by

This is minor **version bump** because it changes functionality in a backward compatible manner

### CODE REVIEW

File: mutacc/utils/vcf_handler.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
+"Module to handle vcf files"
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: Consider making it a separate module

File: mutacc/utils/vcf_handler.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+        results = []
+        for parser in self.parsers:
+            if vcf_entry.get(parser["id"]):
+                info_id = parser["id"]
+                parser_func = parser["parse_func"]
+                results.append(parser_func(vcf_entry[info_id]))
+        return ";".join(results)
+
+    def _get_parsers(self, parse_info):
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: Consider naming and generalization depending on stream direction and what to call the ID

File: mutacc/utils/vcf_handler.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+            given parses an id in the INFO column of the vcf.
+
+            args:
+                entry (dict): dictionary with instructions on how id is parsed
+            returns:
+                parser_func (function): Function that parse ID in INFO column
+        """
+
+        def _type_conv(type_str=None):
+            if type_str == "str":
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: use dispatch table

File: mutacc/utils/vcf_handler.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+            return lambda value: value
+
+        def _parser_func(raw_value):
+            if entry["multivalue"]:
+                info_list = []
+                for raw_value_entry in raw_value.split(entry["separator"]):
+                    element = None
+                    if entry.get("format_separator"):
+                        info_dict = {}
+                        for target, value in zip(
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: Create dict first then loop

File: mutacc/utils/vcf_handler.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                return lambda value: float(value)
+            return lambda value: value
+
+        def _parser_func(raw_value):
+            if entry["multivalue"]:
+                info_list = []
+                for raw_value_entry in raw_value.split(entry["separator"]):
+                    element = None
+                    if entry.get("format_separator"):
+                        info_dict = {}
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: Make a get_info_dict def

File: mutacc/utils/vcf_handler.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                return lambda value: int(value)
+            if type_str == "list":
+                return lambda value: list(value)
+            if type_str == "float":
+                return lambda value: float(value)
+            return lambda value: value
+
+        def _parser_func(raw_value):
+            if entry["multivalue"]:
+                info_list = []
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: parse_multi_value def

File: mutacc/utils/vcf_handler.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+                            raw_value_entry.split(entry["format_separator"]),
+                        ):
+                            target = target.strip()
+                            if entry["target"] == "all" or target in entry["target"]:
+                                info_dict[target] = value
+                        element = info_dict
+                    else:
+                        element = raw_value_entry
+                    info_list.append(element)
+                return _type_conv(entry.get("out_type"))(info_list)
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: Split into more lines

File: mutacc/utils/vcf_handler.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+            else:
+                if entry.get("format") and entry.get("format_separator"):
+                    for target, value in zip(
+                        entry["format"].split(entry["format_separator"]),
+                        raw_value.split(entry["format_separator"]),
+                    ):
+                        target = target.strip()
+                        if target in entry["target"]:
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: should all be here as well?

File: mutacc/utils/vcf_handler.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+            LOG.warning("The stream must be set to either 'read' or 'write' ")
+        self.stream = stream.lower()[0]
+
+        if not self._check(parser_info):
+            LOG.warning("Parser info not given correctly")
+            raise ValueError
+
+        self.parsers = self._get_parsers(parser_info)
+
+    def parse(self, vcf_entry):
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]:
```suggestion
    def parse(self, record):
```

File: mutacc/utils/vcf_handler.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+
+            args:
+                vcf_entry (Cyvcf2.Variant or dict)
+            returns:
+                results (dict): Dictionary with the parsed fields
+        """
+
+        if self.stream == "r":
+            results = {}
+            for parser in self.parsers:
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: read_vcf_record def

File: mutacc/utils/vcf_handler.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+                if vcf_entry.INFO.get(parser["id"]):
+                    info_id = parser["id"]
+                    display_name = parser["display_name"]
+                    parser_func = parser["parse_func"]
+                    results[display_name] = parser_func(vcf_entry.INFO.get(info_id))
+            return results
+
+        results = []
+        for parser in self.parsers:
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: read_mutacc_json_record def

File: mutacc/utils/vcf_handler.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+            if vcf_entry.get(parser["id"]):
+                info_id = parser["id"]
+                parser_func = parser["parse_func"]
+                results.append(parser_func(vcf_entry[info_id]))
+        return ";".join(results)
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]:
```suggestion
        return SEMICOLON.join(results)
```

File: mutacc/utils/vcf_handler.py [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+                            return _type_conv(entry.get("out_type"))(value)
+                else:
+                    return _type_conv(entry.get("out_type"))(raw_value)
+
+        return _parser_func
+
+    @staticmethod
+    def _construct_writer(entry):
+        """
+            Constructs python function that based in key, value pairs in a dictionary
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: Is the doc string correct?

File: mutacc/utils/vcf_handler.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+            creates a INFO entry string in the vcf format
+
+            args:
+                entry (dict): dictionary with instructions on how id is parsed
+            returns:
+                parser_func (function): Function that parse ID in INFO column
+
+        """
+
+        def _writer_func(raw_value):
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: can we pass entry as well

File: mutacc/utils/vcf_handler.py [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+            info_list = []
+            if not isinstance(raw_value, (list, tuple)):
+                raw_value = [raw_value]
+            for element in raw_value:
+                if (
+                    entry.get("target")
+                    and entry.get("format_separator")
+                    and isinstance(element, dict)
+                ):
+                    target_list = (
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: consider splitting into multiple lines

File: mutacc/utils/vcf_handler.py [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+                        list(element.keys())
+                        if entry["target"] == "all"
+                        else entry["target"]
+                    )
+                    info_list.append(
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: make it more clear what is happening here

File: mutacc/utils/vcf_handler.py [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+                    "format_separator"
+                ):
+                    info_list.append(
+                        entry["format_separator"].join([str(item) for item in element])
+                    )
+
+                elif isinstance(element, (int, float, str)):
+                    info_list.append(str(element))
+
+            info_entry = f"{entry['out_name']}={','.join(info_list)}"
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: make it into vcf_info_field and name the part accordingly

File: mutacc/cli/root.py [hunk 18]
```diff
@@ -18,72 +19,85 @@
...
+def cli(context, loglevel, config_file, root_dir, demo, vcf_parser):
 
+    coloredlogs.install(level=loglevel)
     LOG.info("Running mutacc")
 
     cli_config = {}
-
     if demo:
-        host = 'localhost'
+        host = "localhost"
```
[the remark this answers is not in the document]
Comment (adrosenbaum) [reply mutacc/cli/root.py]: Will do! but maybe in other PR...

File: mutacc/cli/root.py [hunk 19]
```diff
@@ -18,72 +19,85 @@
...
+        vcf_info_export = cli_config.get("vcf_info_export")
+        root_dir = cli_config.get("root_dir") or root_dir
         if not root_dir:
-            LOG.warning('Please provide a root directory, through option --root-dir or in config_file')
+            LOG.warning(
+                "Please provide a root directory, through option --root-dir or in config_file"
+            )
             context.abort()
 
+    if vcf_parser is not None:
```
[the remark this answers is not in the document]
Comment (adrosenbaum) [reply mutacc/cli/root.py]: I'll fix that!

**Comment by adrosenbaum** [comment]
@henrikstranneheim what do you think of the code refactoring?

**Comment by henrikstranneheim** [comment]
Sorry did not know you were finished, 😄 . I will take a look tomorrow.

File: mutacc/cli/export.py [hunk 20]
```diff
@@ -8,7 +8,7 @@
 
 from mutacc.parse.path_parse import make_dir
 from mutacc.mutaccDB.query import mutacc_query
-from mutacc.utils.vcf_writer import vcf_writer
+from mutacc.utils.vcf_handler import vcf_writer
```
Comment (henrikstranneheim) [inline mutacc/cli/export.py]: 👍

File: mutacc/utils/constants.py [hunk 21]
```diff
@@ -1 +1,27 @@
+# VCF
```
Comment (henrikstranneheim) [inline mutacc/utils/constants.py]: Very nice

File: mutacc/utils/vcf_handler.py [hunk 22]
```diff
@@ -4,6 +4,21 @@
 
 import yaml
 
+from .constants import (
+    SEMICOLON,
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: This is a quality of life thing, but if you order the imports alphabetically - you can just scan to the import you are looking for or is missing (and find duplicates) without having to read the entire block or until you happen on the one that you want.

Comment (henrikstranneheim) [reply mutacc/utils/vcf_handler.py]: This applies to all code that do not requiring a specific order

File: mutacc/utils/vcf_handler.py [hunk 23]
```diff
@@ -24,33 +39,58 @@ def __init__(self, parser_info: list, stream: str):
...
         if self.stream == "r":
-            results = {}
-            for parser in self.parsers:
-                if vcf_entry.INFO.get(parser["id"]):
-                    info_id = parser["id"]
-                    display_name = parser["display_name"]
-                    parser_func = parser["parse_func"]
-                    results[display_name] = parser_func(vcf_entry.INFO.get(info_id))
-            return results
+            return self._read_vcf_record(record)
```
Comment (henrikstranneheim) [inline mutacc/utils/vcf_handler.py]: Much clearer responsibility 👍

File: tests/utils/test_vcf_handler.py [hunk 24]
```diff
@@ -48,7 +54,7 @@ def test_INFOParser_parse_read():
     """
         Test parsing mocked variant entries given different parser info
     """
-    # test multivalue, array of arrays, eg separated with ',' and then  '|'
+    # test multi_value, array of arrays, eg separated with ',' and then  '|'
```
Comment (henrikstranneheim) [inline tests/utils/test_vcf_handler.py]: I often find it useful to use the Given-When-Then syntax to put the test in context. For instance,
"# Given a multi_value with array of arrays when separated with ','"
- defined data and test

"# Then [descrive expected outcome]"
- assert bla bla

File: tests/utils/test_vcf_handler.py [hunk 25]
```diff
@@ -160,3 +162,102 @@ def test_INFOParser_parse_write_list_of_dict():
     }
     parsed_variant = parser.parse(variant_dict)
     assert parsed_variant == "parsed_value=1|2|3,4|5|6"
+
+
+VARIANT1 = {
```
Comment (henrikstranneheim) [inline tests/utils/test_vcf_handler.py]: Could this be a vfc test fixture for reuse in other tests?

**Comment by adrosenbaum** [comment]
Thanks for your remarks @henrikstranneheim. I'll fix them, then I'll merge this PR.

**Comment by henrikstranneheim** [comment]
👍

**Pull request merged**